File: PF2-web1/web1-backend/src/main/java/com/example/demo/bean/Interface.java
package com.example.demo.bean;

import java.io.Serializable;
import java.util.List;

public class Interface extends Shape implements Cloneable, Serializable {
	private static final long serialVersionUID = 1L;
	private int interface_no;	//交互编号
	private String interface_name;	//交互名称
	private String interface_description;	//交互内容
//	private String interface_intentdescription;	//intent交互内容
	private String interface_from;
	private String interface_to;
	private List<Phenomenon> phenomenonList;	//现象列表
//	private List<Phenomenon> intentphenomenonList;	//intent/context现象列表
//	private List<IntentPhenomenon> intentphenomenonList;	//intent/context现象列表
	private int interface_x1;	//位置信息
	private int interface_y1;
	private int interface_x2;
	private int interface_y2;
	private boolean isintent;
	public Interface(Interface item) {
		interface_no = item.interface_no;
		interface_name = item.interface_name;
		interface_description = item.interface_description;
//		interface_intentdescription = item.interface_intentdescription;
		interface_from = item.interface_from;
		interface_to = item.interface_to;
		phenomenonList = item.phenomenonList;
//		intentphenomenonList = item.intentphenomenonList;
		interface_x1 = item.interface_x1;
		interface_y1 = item.interface_y1;
		interface_x2 = item.interface_x2;
		interface_y2 = item.interface_y2;
		isintent = item.isintent;
	}

	public Interface() {
		super();
	}
	public boolean isIsintent() {
		return isintent;
	}
	public void setIsintent(boolean isintent) {
		this.isintent = isintent;
	}
	public int getInterface_no() {
		return interface_no;
	}
	public void setInterface_no(int interface_no) {
		this.interface_no = interface_no;
	}
	public String getInterface_name() {
		return interface_name;
	}
	public void setInterface_name(String interface_name) {
		this.interface_name = interface_name;
	}
	public String getInterface_description() {
		return interface_description;
	}
	public void setInterface_description(String interface_description) {
		this.interface_description = interface_description;
	}

//	public String getInterface_intentdescription() {
//		return interface_intentdescription;
//	}
//
//	public void setInterface_intentdescription(String interface_intentdescription) {
//		this.interface_intentdescription = interface_intentdescription;
//	}
	public String getInterface_from() {
		return interface_from;
	}
	public void setInterface_from(String interface_from) {
		this.interface_from = interface_from;
	}
	public String getInterface_to() {
		return interface_to;
	}
	public void setInterface_to(String interface_to) {
		this.interface_to = interface_to;
	}
	public List<Phenomenon> getPhenomenonList() {
		return phenomenonList;
	}
	public void setPhenomenonList(List<Phenomenon> phenomenonList) {
		this.phenomenonList = phenomenonList;
	}
//	public List<Phenomenon> getIntentPhenomenonList() {
//		return intentphenomenonList;
//	}
//	public void setIntentPhenomenonList(List<Phenomenon> intentPhenomenonList) {
//		this.intentphenomenonList = intentPhenomenonList;
//	}
	public int getInterface_x1() {
		return interface_x1;
	}
	public void setInterface_x1(int interface_x1) {
		this.interface_x1 = interface_x1;
	}
	public int getInterface_y1() {
		return interface_y1;
	}
	public void setInterface_y1(int interface_y1) {
		this.interface_y1 = interface_y1;
	}
	public int getInterface_x2() {
		return interface_x2;
	}
	public void setInterface_x2(int interface_x2) {
		this.interface_x2 = interface_x2;
	}
	public int getInterface_y2() {
		return interface_y2;
	}
	public void setInterface_y2(int interface_y2) {
		this.interface_y2 = interface_y2;
	}
	@Override    
	public Object clone() {        
		Interface inte = null;        
		try {            
			inte = (Interface) super.clone();        
		} catch (CloneNotSupportedException e) {            
			e.printStackTrace();        
			}        
		return inte;    
	}

	@Override
	public int getNo() {
		return interface_no;
	}

	@Override
	public void setNo(int node_no) {
		this.interface_no=node_no;
	}

	@Override
	public String getName() {
		return interface_name;
	}

	@Override
	public void setName(String node_no) {
		this.interface_name=node_no;
	}
	public void refreshPhenomenonList(String oldShortName, String newShortName) {
		for (Phenomenon phenomenon : phenomenonList) {
			if (phenomenon.getPhenomenon_from().equals(oldShortName))
				phenomenon.setPhenomenon_from(newShortName);
			else if (phenomenon.getPhenomenon_to().equals(oldShortName))
				phenomenon.setPhenomenon_to(newShortName);
		}
//		if(intentphenomenonList != null) {
//			for (Phenomenon phenomenon : intentphenomenonList) {
//				if (phenomenon.getPhenomenon_from().equals(oldShortName))
//					phenomenon.setPhenomenon_from(newShortName);
//				else if (phenomenon.getPhenomenon_to().equals(oldShortName))
//					phenomenon.setPhenomenon_to(newShortName);
//			}
//		}
	}

}

File: PF2-web5/web5-vue/web2-vue/src/types/Interface.ts
import { Line } from './Line';
import { Phenomenon } from './Phenomenon';

export class Interface extends Line {
  interface_no: number;
  interface_name: string;
  interface_description: string;
  interface_from: string;
  interface_to: string;
  phenomenonList: Phenomenon[];
  interface_x1: number;
  interface_y1: number;
  interface_x2: number;
  interface_y2: number;

  static newInterface(no: number, name: string, description: string, from: string, to: string, phe: Phenomenon[], x1: number, y1: number, x2: number, y2: number) {
    let int = new Interface();
    int.interface_no = no;
    int.interface_name = name;
    int.interface_description = description;
    int.interface_from = from;
    int.interface_to = to;
    int.phenomenonList = phe;
    int.interface_x1 = x1;
    int.interface_y1 = y1;
    int.interface_x2 = x2;
    int.interface_y2 = y2;
    return int;
  }

  static newInterfaceWithOld(old: Interface, phenomenonList: Phenomenon[], description: string) {
    let int = new Interface();
    int.interface_no = old.interface_no;
    int.interface_name = old.interface_name;
    int.interface_description = description;
    int.interface_from = old.interface_from;
    int.interface_to = old.interface_to;
    int.phenomenonList = phenomenonList;
    int.interface_x1 = old.interface_x1;
    int.interface_y1 = old.interface_y1;
    int.interface_x2 = old.interface_x2;
    int.interface_y2 = old.interface_y2;
    return int;
  }

  static getDescription1(name: string, pheList: Phenomenon[]) {
    // a:M!{on},P!{off}
    let s = "";
    s = s + name + ":";
    let desList: [string, string][] = [];
    for (let phe of pheList) {
      let flag = false;
      for (let des of desList) {
        if (phe.phenomenon_from == des[0]) {
          des.push(phe.phenomenon_name);
          flag = true;
          break;
        }
      }
      if (!flag) {
        desList.push([phe.phenomenon_from, phe.phenomenon_name]);
      }
    }

    for (let des of desList) {
      s += des[0] + "!{";
      for (let item of des.slice(1)) {
        s += item + ",";
      }
      s = s.slice(0, -1);
      s += "},";
    }
    s = s.slice(0, -1);
    return s;
  }

  static copyInterface(old: Interface) {
    let int = new Interface();
    int.interface_no = old.interface_no;
    int.interface_name = old.interface_name;
    int.interface_from = old.interface_from;
    int.interface_to = old.interface_to;
    if (old.phenomenonList != undefined) {
      int.phenomenonList = old.phenomenonList;
      int.interface_description = Interface.getDescription1(int.interface_name, old.phenomenonList);
    } else {
      int.phenomenonList = new Array<Phenomenon>();
    }
    int.interface_description = Interface.getDescription1(int.interface_name, int.phenomenonList);
    int.interface_x1 = old.interface_x1;
    int.interface_y1 = old.interface_y1;
    int.interface_x2 = old.interface_x2;
    int.interface_y2 = old.interface_y2;
    return int;
  }

  getNo() { return this.interface_no; }
  setNo(no: number) { this.interface_no = no; }

  getName() { return this.interface_name; }
  setName(name: string) { this.interface_name = name; }

  getDescription() { return this.interface_description; }
}
